### PR TITLE
Make the experimental route /metrics activable via HTTP

### LINKED ISSUE #4086: Make the experimental route `/metrics` activable via HTTP too
We would like to make the experimental “Prometheus metrics” feature accessible via the HTTP API, as it is for the [score details](https://github.com/meilisearch/product/discussions/674) and [vector search](https://github.com/meilisearch/product/discussions/677) experimental features.

More info regarding this exp feature: https://github.com/meilisearch/product/discussions/625

### Activate the feature:

The experimental feature to activate the route `/metrics` is currently activable using a flag `--experimental-enable-metrics`

```jsx
curl \
  -X PATCH 'http://localhost:7700/experimental-features/' \
  -H 'Content-Type: application/json'  \
--data-binary '{
    "metrics": true
  }'
```

### Getting the information about the feature

```jsx
curl -X GET 'http://localhost:7700/experimental-features/'

{
    "scoreDetails": false,
    "vectorStore": false,
    "metrics": true
}
```

## TODO

- [ ]  Make it available via HTTP as described above
- [ ]  The users can still launch Meilisearch using the `--experimental-enable-metrics` flag.
- [ ]  If the flag `--experimental-enable-metrics` is activated, a call to the `GET /experimental-features` route right after the launch will show `"metrics": true` even if the user has not called the `PATCH /experimental-features` route yet.
- [ ]  Even if the `--experimental-enable-metrics` flag is present at launch, calling the `PATCH /experimental-features` route with `"metrics": false` disables the experimental feature.
- [ ]  Update the spec

## Impacted teams
@meilisearch/cloud-team @meilisearch/docs-team

### PR DESCRIPTION
**Pull request by braddotcoffee**
# Pull Request

## Related issue
Closes #4086

## What does this PR do?
- [x] Make `/metrics` available via HTTP as described in #4086 
- [x] The users can still launch Meilisearch using the `--experimental-enable-metrics` flag.
- [x] If the flag `--experimental-enable-metrics` is activated, a call to the `GET /experimental-features` route right after the launch will show `"metrics": true` even if the user has not called the `PATCH /experimental-features` route yet.
- [x] Even if the --experimental-enable-metrics flag is present at launch, calling the `PATCH /experimental-features` route with `"metrics": false` disables the experimental feature.
- [x] Update the spec
    - I was unable to find docs in this repository to update about the `/experimental-features` endpoint. I'll happily update if you point me in the right direction!

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

### CODE REVIEW

**Review by curquiza — COMMENTED** [review]
The PR is functional, I did not review the code however
Thank you very much @braddotcoffee 

For the spec: I should have precise it, it's not for external contribution, we will update ourself 😊

**Review by Kerollmops — DISMISSED** [review]
Looks good to me, that's working quite well ☕
Thank you.

@vivek-26, would it be possible to review it too when you have time? I, maybe, missed something important.

**Comment by Kerollmops** [comment]
> The metrics middleware needs to be updated [here](https://github.com/meilisearch/meilisearch/blob/main/meilisearch/src/lib.rs#L117-L120) because it can be toggled on or off at runtime. Currently, it is enabled only if we pass `--experimental-enable-metrics` flag.

Indeed, you are right. The middleware should always be enabled and conditionally ignored at runtime instead by using the `IndexScheduler::metrics_enabled` method.

File: index-scheduler/src/features.rs [hunk 1]
```diff
@@ -73,9 +71,12 @@ impl FeatureData {
...
         let mut wtxn = env.write_txn()?;
         let runtime_features = env.create_database(&mut wtxn, Some(EXPERIMENTAL_FEATURES))?;
+        let default_features =
+            RuntimeTogglableFeatures { metrics: instance_features.metrics, ..Default::default() };
+        runtime_features.put(&mut wtxn, EXPERIMENTAL_FEATURES, &default_features)?;
         wtxn.commit()?;
 
-        Ok(Self { runtime: runtime_features, instance: instance_features })
+        Ok(Self { runtime: runtime_features })
     }
```
[the remark this answers is not in the document]
Comment (Kerollmops) [reply index-scheduler/src/features.rs:80]: You are right! Good catch! However, after reviewing the different behaviors, it is likely our best behavior when combining the CLI arg and the runtime HTTP route. I would like to document and advice to always use the HTTP route and avoid using the CLI argument.

**Comment by curquiza** [comment]
@braddotcoffee can you update your PR according to the review?
Also, looks like they are git conflicts, can you fix them? 😊

**Comment by braddotcoffee** [comment]
Thanks for the review! I'll get to work on this!

File: index-scheduler/src/features.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (braddotcoffee) [reply index-scheduler/src/features.rs:80]: Just wanted to check to make sure I'm understanding expected behavior correctly while working on updating this PR.

I just ran the following test:
#### Step One: Run server with `--experimental-enable-metrics`
```bash
cargo +nightly run -- --experimental-enable-metrics
```

#### Step Two: Check experimental features via HTTP
```bash
curl -X GET 'http://localhost:7700/experimental-features/'

{
  "scoreDetails": false,
  "vectorStore": false,
  "metrics": true
}
```

#### Step Three: Run server without `--experimental-enable-metrics`
```bash
cargo +nightly run
```

#### Step Four: Check experimental features via HTTP
```bash
curl -X GET 'http://localhost:7700/experimental-features/'

{
  "scoreDetails": false,
  "vectorStore": false,
  "metrics": false
}
```
---
Is this the expected behavior? Or am I missing something about my testing that would cause different behavior?

**Comment by braddotcoffee** [comment]
I could not find a place where the current `RouteMetricsMiddleware` was being tested. I did perform manual tests to make sure that the timing logic was being executed as expected, but I can understand this is brittle.

I attempted to write a test that would verify the middleware behavior by checking `MEILISEARCH_HTTP_REQUESTS_TOTAL`, but as (I assume) the tests are multithreaded, this proved to be flaky. Do you have any thoughts on a more reliable automated test I could add for peace of mind?

File: meilisearch/src/middleware.rs [hunk 2]
```diff
@@ -47,19 +61,21 @@ where
 
     fn call(&self, req: ServiceRequest) -> Self::Future {
         let mut histogram_timer: Option<HistogramTimer> = None;
-        let request_path = req.path();
-        let is_registered_resource = req.resource_map().has_resource(request_path);
-        if is_registered_resource {
-            let request_method = req.method().to_string();
-            histogram_timer = Some(
-                crate::metrics::MEILISEARCH_HTTP_RESPONSE_TIME_SECONDS
+        if let Ok(()) = self.index_scheduler.features().and_then(|features| features.check_metrics()) {
```
[the remark this answers is not in the document]
Comment (Kerollmops) [reply meilisearch/src/middleware.rs:64]: If I am correct, the `index_scheduler.features()` call is requesting a read transaction and doing a single read in LMDB, right? It doesn't cost much as we will do a lot of reads during the search and do one or two when we need to fetch the requested index, anyway. No performance hit should be seen.

File: meilisearch/src/middleware.rs [hunk 3]
```diff
@@ -4,15 +4,28 @@ use std::future::{ready, Ready};
...
 
+pub struct RouteMetricsMiddlewareFactory {
+    index_scheduler: Data<IndexScheduler>,
+}
+
+impl RouteMetricsMiddlewareFactory {
+    pub fn new(index_scheduler: Data<IndexScheduler>) -> Self {
+        RouteMetricsMiddlewareFactory { index_scheduler }
+    }
+}
```
[the remark this answers is not in the document]
Comment (braddotcoffee) [reply meilisearch/src/middleware.rs:22]: Sorry about that - new to actix and didn't realize I'd have access to this without passing it into the Middleware. Making the update now!

File: meilisearch/src/middleware.rs [hunk 4]
```diff
@@ -47,19 +49,29 @@ where
...
-                crate::metrics::MEILISEARCH_HTTP_RESPONSE_TIME_SECONDS
+        let data = req.app_data::<Data<IndexScheduler>>();
+        let metrics_enabled = data
+            .ok_or("Could not get index scheduler")
+            .and_then(|index_scheduler| {
+                index_scheduler.features().map_err(|_| "Could not get features")
+            })
+            .and_then(|features| features.check_metrics().map_err(|_| "Metrics not enabled"));
+
+        if let Ok(()) = metrics_enabled {
```
[the remark this answers is not in the document]
Comment (Kerollmops) [reply meilisearch/src/middleware.rs:60]: It would indeed be better to return an error. It would be much better than the current strings not being displayed in case of error here. Would it be possible to return a 500 with the error message? Even an ugly internal one?

File: index-scheduler/src/features.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kerollmops) [reply index-scheduler/src/features.rs:93]: Thank you for this information. Does it mean not specifying the flag will always put `"metrics": false` in the database?

File: meilisearch/src/middleware.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (braddotcoffee) [reply meilisearch/src/middleware.rs:60]: Will update! I was worried about this middleware failure taking down the ability to handle any requests so I didn't want to return the error off the bat. I should've left a comment asking about expected behavior there. I can see that you wouldn't want metrics to disappear silently

File: index-scheduler/src/features.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dureuill) [reply index-scheduler/src/features.rs:93]: Hello @vivek-26, @braddotcoffee, @Kerollmops  :wave:

The complexity here is that we have two states to handle:

1. The "immediate" state of the feature, is it currently enabled or not? ("Feature state")
2. The "persistent" state stored in the DB ("DB state").

And two ways to act on the state:

1. Setting or not the CLI flag
2. Calling the experimental route to enable/disable the feature.

Taking this into account, I feel like the most intuitive behavior would be:

| DB state before operation | Action | Feature state after action | DB state after action | Note |
|---------------------------|--------|--------------------------|----------------------|------|
| ❌ | Start without CLI flag | ❌ | ❌ | |
| ❌ | Start with CLI flag | ✅ | ❌ | CLI flag enable features, doesn't change DB state |
| ❌ | Enable by calling the route | ✅ | ✅ | Regardless of the presence or absence of the CLI flag at startup |
| ❌ | Disable by calling the route | ❌ | ❌ | Regardless of the presence or absence of the CLI flag at startup |
| ✅ | Start without CLI flag | ✅ | ✅ | Lack of CLI flag never disables the feature, doesn't change DB state |
| ✅ | Start without CLI flag | ✅ | ✅ | |
| ✅ | Enable by calling the route | ✅ | ✅ | Regardless of the presence or absence of the CLI flag at startup |
| ✅ | Disable by calling the route | ❌ | ❌ | Regardless of the presence or absence of the CLI flag at startup |

Like @vivek-26 implied, setting or not setting the CLI flag should never affect the DB state, only the "immediate" feature state. Furthermore, not setting the CLI flag should not disable the feature, in that case the DB state should take precedence.

In short `initialState = db.state || cli.state`

[the remark this answers is not in the document]
Comment (braddotcoffee) [reply index-scheduler/src/features.rs:93]: I can happily update to this! Just wanna make sure everyone is on the same page about it. @vivek-26 had mentioned in the original comment that this behavior might be unintuitive:
> let's say initially (when the server starts), at runtime level, metrics is false in DB, and, we pass `--experimental-enable-metrics` which gets persisted to DB.
>
> Now, if the user never updates the metrics feature using `PATCH /experimental-features`, **the next time if the server runs without `--experimental-enable-metrics`, the metrics feature will be enabled (as it was persisted to DB the last time server had started), which is incorrect I feel.**

This seems to be the behavior described by this line of your table:
| DB state before operation  | Action | Feature State After Action | DB State After Action | Note | 
| --- | --- | --- | --- | --- |
| ✅ | Start without CLI flag | ✅ | ✅ | Lack of CLI flag never disables the feature, doesn't change DB state |

@vivek-26 / @Kerollmops are you in alignment with @dureuill's table?

File: index-scheduler/src/features.rs [hunk 5]
```diff
@@ -85,27 +87,38 @@ impl RoFeatures {
...
         &self,
         mut wtxn: RwTxn,
         features: RuntimeTogglableFeatures,
     ) -> Result<()> {
-        self.runtime.put(&mut wtxn, EXPERIMENTAL_FEATURES, &features)?;
+        self.persisted.put(&mut wtxn, EXPERIMENTAL_FEATURES, &features)?;
         wtxn.commit()?;
+
+        let mut toggled_features = self.runtime.write().map_err(|_| RuntimeFeatureToggleError)?;
+        *toggled_features = features;
```
Comment (braddotcoffee) [inline index-scheduler/src/features.rs:113]: It is technically possible to end up in a weird state here where the runtime representation is different from what's persisted (your db write succeeds but your `RwLock` write fails). I did not think this would cause major problems due to the ability to retry the `put_runtime_features` that would rectify this discrepancy but I can handle this differently if you want!

**Review by dureuill — APPROVED** [review]
Thank you a lot for your contribution, and for being open to feedback.

I took the liberty to make some changes to your PR to simplify error handling around the `RwLock`, which cascaded to making `index_scheduler.features()` non-fallible. See inline comments for details and rationale.

Thanks again for your contribution :heart:, with these changes I believe the PR is ready to land :rocket:

**Comment by dureuill** [comment]
bors merge
